File: node/simple.js
const fs = require("fs");
const protoLoader = require("@grpc/proto-loader");
const grpc = require("grpc");
const wav = require("wav");

const appId = process.env.APP_ID;
if (appId === undefined) {
  throw new Error("APP_ID environment variable needs to be set");
}

let host = "api.speechgrinder.com";
let credentials = grpc.credentials.createSsl();

const SgGrpc = grpc.loadPackageDefinition(
  protoLoader.loadSync("../sg.proto", {
    keepCase: false,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true
  })
);

const login = (deviceId, appId) => {
  return new Promise((resolve, reject) => {
    const identity = new SgGrpc.speechgrinder.sgapi.v1.Identity(host, credentials);
    identity.login({ appId, deviceId }, (err, response) => {
      if (err) {
        return reject(err);
      }
      return resolve(response.token);
    });
  });
};

const start = slu => {
  const audio = new wav.Reader();

  slu.write({ config: { channels: 1, sampleRateHertz: 16000 } });
  slu.write({ event: { event: "START" } });

  audio.on("data", audioData => {
    if (slu.writable) {
      slu.write({ audio: audioData });
    }
  });
  slu.on("data", data => {
    console.dir(data, { depth: null });
  });
  audio.on("end", () => {
    slu.write({ event: { event: "STOP" } });
    slu.end();
  });
  fs.createReadStream("../audio.wav").pipe(audio);
};

Promise.resolve()
  .then(() => login("node-simple-test", appId))
  .catch(err => {
    console.error(err);
    process.exit();
  })
  .then(token => {
    const metadata = new grpc.Metadata();
    metadata.add("Authorization", `Bearer ${token}`);
    const client = new SgGrpc.speechgrinder.sgapi.v1.Slu(host, credentials);
    const slu = client.Stream(metadata);
    return start(slu);
  })
  .catch(err => {
    console.error(err);
  });

File: python/simple.py
import os
import wave

import grpc

from sg_pb2 import SluRequest, SluConfig, SluEvent, LoginRequest
from sg_pb2_grpc import IdentityStub as IdentityService
from sg_pb2_grpc import SluStub as SLUService

chunk_size = 8000

if 'APP_ID' not in os.environ:
    raise RuntimeError('APP_ID environment variable needs to be set')

def audio_iterator():
    with wave.open('../audio.wav', mode='r') as audio_file:
        if audio_file.getnchannels() > 1:
            raise RuntimeError('Only mono channel audio is supported')
        yield SluRequest(config=SluConfig(channels=1, sample_rate_hertz=audio_file.getframerate()))
        yield SluRequest(event=SluEvent(event='START'))
        audio_bytes = audio_file.readframes(chunk_size)
        while audio_bytes:
            yield SluRequest(audio=audio_bytes)
            audio_bytes = audio_file.readframes(chunk_size)
    yield SluRequest(event=SluEvent(event='STOP'))


with grpc.secure_channel('api.speechgrinder.com', grpc.ssl_channel_credentials()) as channel:
    token = IdentityService(channel)\
        .Login(LoginRequest(device_id='python-simple-test', app_id=os.environ['APP_ID']))\
        .token

with grpc.secure_channel('api.speechgrinder.com', grpc.ssl_channel_credentials()) as channel:
    slu = SLUService(channel)
    responses = slu.Stream(
        audio_iterator(),
        None,
        [('authorization', 'Bearer {}'.format(token))])
    for response in responses:
        print(response)
